### PR TITLE
windows: *.bat and *.cmd files to search

### PR DESCRIPTION
**Pull request by dnakamura**
Fixes joyent/node#25330

### CODE REVIEW

**Comment by piscisaureus** [comment]
This is a good idea. However I suspect that when CreateProcess() runs .bat/.cmd files, the arguments aren't passed correctly. Can you check this?

Ideally I'd like to see a test that spawns `test.bat` with a list of complicated arguments, e.g. ['hello', 'world^', '"$ &"'], and verifies that the .bat/.cmd file receives these arguments correctly.

**Comment by dnakamura** [comment]
I did a simple test and it did seem to pass everything correctly. However doing a little digging it seems that while CreateProcess() happily launches .bat/.cmd file the docs state that you need to launch run with cmd.exe /c script.bat

**Comment by mhdawson** [comment]
A few minor comments:

1) Missing space between ){ in an few places

2) I believe we should be using uv__malloc/uv__free instead of malloc/free, see other places allocation is done in the file

3) Based on the suggestions from @piscisaureus  I think you need additional tests to cover the suggested cases ['hello', 'world^', '"$ &"']

**Comment by dnakamura** [comment]
@piscisaureus  Can you take a look at the new changes? I've added the tests, and incorporated the changes @mhdawson suggested.

**Comment by piscisaureus** [comment]
I can still think of a few edge cases. Let me get back to this in a bit.

File: src/win/process.c [hunk 1]
```diff
@@ -250,6 +250,30 @@ static WCHAR* search_path_join_test(const WCHAR* dir,
   return NULL;
 }
 
+/*
+ * Gets a string which contains the full path to cmd.exe
+ * caller responsible for freeing memory on success
```
Comment (piscisaureus) [inline src/win/process.c]: This should either be an internal-only API, or the user shouldn't be responsible for allocating the buffer.
The reason is that the user `free()` function may not work with the allocator used for `uv__alloc()`.
cc @bnoordhuis - you agree?

File: src/win/process.c [hunk 2]
```diff
@@ -289,6 +315,27 @@ static WCHAR* path_search_walk_ext(const WCHAR *dir,
                                  L"exe", 3,
                                  cwd, cwd_len);
   if (result != NULL) {
+    *isScript = FALSE;
+    return result;
+  }
+
+  /* Try .cmd extension */
```
Comment (piscisaureus) [inline src/win/process.c]: This is fine for now but eventually it should honor the `PATHEXT` environment variable.

File: src/win/process.c [hunk 3]
```diff
@@ -375,6 +423,9 @@ static WCHAR* search_path(const WCHAR *file,
   /* Check if the filename includes an extension */
   dot = wcschr(file_name_start, L'.');
   name_has_ext = (dot != NULL && dot[1] != L'\0');
+  if (name_has_ext) {
+    *isScript = (_wcsicmp(dot, L".bat") == 0 )||(_wcsicmp(dot, L".cmd") == 0);
```
Comment (piscisaureus) [inline src/win/process.c]: * Style: add spaces around the `||`.
* I think this would trigger a false positive if the file name had .bat somewhere int he path, e.g. `c:\foo\bert.bat\run.exe`.

File: src/win/process.c [hunk 4]
```diff
@@ -569,7 +631,20 @@ int make_program_args(char** args, int verbatim_arguments, WCHAR** dst_ptr) {
...
 
   pos = dst;
-  for (arg = args; *arg; arg++) {
+  arg = args;
+  /* If launching script file, add 'cmd /C scriptname',
+     and ignore first arg */
+  if (script_file != NULL) {
+    /* advance arg pointer if args[0] was set properly */
+    if(*arg != NULL) arg++;
+    wcscpy(pos, L"cmd /c");
```
Comment (piscisaureus) [inline src/win/process.c]: Unfortunately `cmd /c` behaves weirdly if the command line starts and ends with a quote (e.g. `"c:\program files\bar.bat" 1 2 "hello world"`. We would need a way to work around that.

File: test/test-spawn.c [hunk 5]
```diff
@@ -1079,6 +1083,60 @@ TEST_IMPL(argument_escaping) {
   return 0;
 }
 
+TEST_IMPL(spawn_batch_args) {
+  int r, i;
+  char* test_args[] = {
```
Comment (piscisaureus) [inline test/test-spawn.c]: I think a good test would actually *spawn* a .batch file and verifies that the arguments come through correctly. The reason is that nobody really understand cmd's command line parsing.
A way to do this is by having the batch file print all arguments to stdout on a separate line, and verifies that the output is what you expected.

```
@echo off
:loop
if "%1"=="" goto :eof
echo %1
shift
goto loop
```

File: src/win/process.c (hunk 3 above)
Comment (dnakamura) [inline src/win/process.c]: @piscisaureus no this code is fine since it is starting from filename component of the original string. However since it serarches for the first occurence of ".", it could get thrown off if you have for example ```c:\some\dir\script.blah.bat```

**Comment by szarouski** [comment]
Hi, sorry I can't contribute to fix this, but I wonder if this is still on the radar. Thanks.

**Comment by saghul** [comment]
@szarouski I'm afraid it fell through the cracks :-S

**Comment by szarouski** [comment]
@saghul, as a temp workaround I use https://www.npmjs.com/package/npm-which so it finds a file with .bat, .cmd extension for me.

**Comment by bzoz** [comment]
Like with https://github.com/libuv/libuv/pull/392 - `cmd.exe` should be used to spawn scripts.

/cc @bnoordhuis

**Comment by bnoordhuis** [comment]
Right you are, I'll close this out.  Thanks anyway, @dnakamura.
